File: clients/python/examples/sigfirst.py
"""Subscribe to sig-first with an explicit account or program filter.

PULSE_TARGET='<HOST:PORT_FROM_DASHBOARD>' \
PULSE_TOKEN='<TOKEN_FROM_SAME_LOCATION>' \
PULSE_ACCOUNT='<ACCOUNT_OR_PROGRAM_PUBKEY>' python examples/sigfirst.py
"""

import asyncio
import os

import base58

from thornode_pulse import Filter, connect_pulse


async def main() -> None:
    target = os.environ["PULSE_TARGET"]
    token = os.environ["PULSE_TOKEN"]
    account_or_program = os.environ["PULSE_ACCOUNT"]
    flt = Filter.accounts(account_or_program)

    async with connect_pulse(target, token=token) as client:
        sub = await client.subscribe_sig_first(flt)
        async for item in sub:
            signature = base58.b58encode(item.signature).decode("ascii")
            print(f"slot={item.slot} seq={item.seq} signature={signature}")
            print(f"metrics={sub.metrics} gaps={sub.gaps}")


if __name__ == "__main__":
    asyncio.run(main())

File: clients/rust/examples/sig_first.rs
//! Subscribe to the sig-first tier and print `(slot, signature)` as they arrive.
//!
//!   PULSE_ADDR='<HOST:PORT_FROM_DASHBOARD>' PULSE_TOKEN='<TOKEN_FROM_SAME_LOCATION>' PULSE_ACCOUNT='<ACCOUNT_OR_PROGRAM_PUBKEY>' cargo run -p thornode-pulse --example sig_first

use thornode_pulse::Filter;

mod common;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let addr = std::env::var("PULSE_ADDR").expect("set PULSE_ADDR to host:port from dashboard");
    let token =
        std::env::var("PULSE_TOKEN").expect("set PULSE_TOKEN from the same dashboard location");
    let client = common::connect(&addr, Some(token)).await?;
    let account =
        std::env::var("PULSE_ACCOUNT").expect("set PULSE_ACCOUNT to an account or program pubkey");
    println!("connected to {addr}; subscribing sig-first…");

    let mut sub = client
        .subscribe_sig_first(&Filter::accounts([account]))
        .await?;

    let mut n: u64 = 0;
    while let Some(item) = sub.next().await? {
        println!(
            "slot {:>12}  seq={}  {}",
            item.slot,
            item.seq,
            bs58::encode(item.signature).into_string()
        );
        n += 1;
        if n >= 20 {
            break;
        }
    }
    println!(
        "done ({n} txs); dropped={} gaps={}",
        sub.dropped(),
        sub.gaps()
    );
    Ok(())
}
